File: src/main/scala/temportalist/compression/main/client/model/ItemListCompressed.scala
package temportalist.compression.main.client.model

import java.util

import com.google.common.collect.Lists
import net.minecraft.block.state.IBlockState
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.block.model.SimpleBakedModel.Builder
import net.minecraft.client.renderer.block.model._
import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraft.entity.EntityLivingBase
import net.minecraft.item.{ItemBlock, ItemStack}
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import temportalist.compression.main.common.Compression
import temportalist.compression.main.common.init.Compressed
import temportalist.compression.main.common.lib.EnumTier
import temportalist.origin.api.common.helper.Names

/**
  *
  * Created by TheTemportalist on 4/14/2016.
  *
  * @author TheTemportalist
  */
class ItemListCompressed(private val overlays: Array[TextureAtlasSprite])
		extends ItemOverrideList(Lists.newArrayList()) {

	def getMissingModel: IBakedModel =
		Minecraft.getMinecraft.getBlockRendererDispatcher.getBlockModelShapes.getModelManager.getMissingModel

	override def handleItemState(originalModel: IBakedModel, stack: ItemStack, world: World,
			entity: EntityLivingBase): IBakedModel = {
		if (!stack.hasTagCompound) return this.getMissingModel

		val sampleStack = Compressed.getSampleStack(stack)
		val isBlock = stack.getItem.isInstanceOf[ItemBlock]

		val sampleModel =
			if (isBlock)
				Minecraft.getMinecraft.getBlockRendererDispatcher.getBlockModelShapes.
						getModelForState(Compressed.getSampleState(stack))
			else
				Minecraft.getMinecraft.getRenderItem.getItemModelMesher.getItemModel(sampleStack)

		val size = Compressed.getSize(stack)
		val i = EnumTier.getTierForSize(size).ordinal()
		val overlay = overlays(i)

		new IBakedModel {

			override def getParticleTexture: TextureAtlasSprite = sampleModel.getParticleTexture

			override def isBuiltInRenderer: Boolean = sampleModel.isBuiltInRenderer

			override def getItemCameraTransforms: ItemCameraTransforms = sampleModel.getItemCameraTransforms

			override def isAmbientOcclusion: Boolean = true

			override def isGui3d: Boolean = isBlock

			override def getOverrides: ItemOverrideList = sampleModel.getOverrides

			override def getQuads(state: IBlockState,
					side: EnumFacing,
					rand: Long): util.List[BakedQuad] = {

				val quadList = new util.ArrayList[BakedQuad]()

				try {
					quadList.addAll(sampleModel.getQuads(state, side, rand))
					if (overlay != null) {
						val overlayModel = new Builder(
							state, sampleModel, overlay, BlockPos.ORIGIN).makeBakedModel()
						quadList.addAll(overlayModel.getQuads(state, side, rand))
					}
				}
				catch {
					case e: Exception =>
						Compression.log("Error merging render models. " +
								"Please report this to https://github.com/TheTemportalist/Compression/issues. " +
								"As a temporary fix, you can consider adding \'" +
								Names.getName(stack, hasID = true, hasMeta = false) +
								"\' or \'" +
								Names.getName(stack, hasID = true, hasMeta = true) +
								"\' to the blacklist configuration option.")
						e.printStackTrace()
				}

				quadList
			}

		}
	}

}

File: src/main/java/com/temportalist/compression/client/ItemListCompressed.java
package com.temportalist.compression.client;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Lists;
import com.temportalist.compression.common.Compression;
import com.temportalist.compression.common.init.CompressedStack;
import com.temportalist.compression.common.lib.EnumTier;
import net.minecraft.block.state.IBlockState;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.block.model.BakedQuad;
import net.minecraft.client.renderer.block.model.IBakedModel;
import net.minecraft.client.renderer.block.model.ItemCameraTransforms;
import net.minecraft.client.renderer.block.model.ItemOverrideList;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

import javax.annotation.Nullable;
import javax.vecmath.Matrix4f;
import java.util.ArrayList;
import java.util.List;
import net.minecraft.client.renderer.block.model.SimpleBakedModel.Builder;
import net.minecraftforge.client.model.PerspectiveMapWrapper;
import net.minecraftforge.common.model.TRSRTransformation;
import org.apache.commons.lang3.tuple.Pair;

public class ItemListCompressed extends ItemOverrideList {

    TextureAtlasSprite[] overlays;

    public ItemListCompressed(TextureAtlasSprite[] overlays) {
        super(Lists.newArrayList());
        this.overlays = overlays;
    }

    public static IBakedModel getMissingModel() {
        return Minecraft.getMinecraft().getBlockRendererDispatcher().getBlockModelShapes().getModelManager().getMissingModel();
    }

    @Override
    public IBakedModel handleItemState(IBakedModel originalModel, ItemStack stack, @Nullable World world, @Nullable EntityLivingBase entity) {
        if (!stack.hasTagCompound()) return originalModel;

        ItemStack sampleStack = CompressedStack.createSampleStack(stack);
        boolean isBlock = stack.getItem() instanceof ItemBlock;

        IBakedModel sampleModel = isBlock ?
                Minecraft.getMinecraft().getBlockRendererDispatcher().getBlockModelShapes().getModelForState(CompressedStack.createSampleState(stack)) :
                Minecraft.getMinecraft().getRenderItem().getItemModelMesher().getItemModel(sampleStack);

        int i = CompressedStack.getTier(stack).ordinal();
        TextureAtlasSprite overlay = overlays[i];

        return new IBakedModel() {

            @Override
            public TextureAtlasSprite getParticleTexture() {
                return sampleModel.getParticleTexture();
            }

            @Override
            public boolean isBuiltInRenderer() {
                return false;
            }

            @Override
            public boolean isAmbientOcclusion() {
                return true;
            }

            @Override
            public boolean isGui3d() {
                return isBlock;
            }

            @Override
            public ItemOverrideList getOverrides() {
                return sampleModel.getOverrides();
            }

            @Override
            public List<BakedQuad> getQuads(@Nullable IBlockState state, @Nullable EnumFacing side, long rand) {
                ArrayList<BakedQuad> quadList = new ArrayList<>();

                try {
                    List<BakedQuad> quads = sampleModel.getQuads(state, side, rand);
                    if (quads != null) quadList.addAll(quads);
                    if (overlay != null) {
                        IBakedModel overlayModel = new Builder(state, sampleModel, overlay, BlockPos.ORIGIN).makeBakedModel();
                        quads = overlayModel.getQuads(state, side, rand);
                        if (quads != null) quadList.addAll(quads);
                    }
                }
                catch(Exception e) {
                    Compression.LOGGER.error("Error merging render models. " +
                            "Please report this to https://github.com/TheTemportalist/Compression/issues. " +
                            "As a temporary fix, you can consider adding \'" +
                            (sampleStack.getItem().getRegistryName().toString()) +
                            "\' or \'" +
                            (sampleStack.getItem().getRegistryName().toString() + ":" + sampleStack.getItemDamage()) +
                            "\' to the blacklist configuration option.");
                    e.printStackTrace();
                }

                return quadList;
            }

            @Override
            public Pair<? extends IBakedModel, Matrix4f> handlePerspective(ItemCameraTransforms.TransformType cameraTransformType) {
                ImmutableMap<ItemCameraTransforms.TransformType, TRSRTransformation> map =
                        isBlock ? TransformHelper.BLOCK : TransformHelper.ITEM;
                return PerspectiveMapWrapper.handlePerspective(
                        this, map, cameraTransformType
                );
            }

        };
    }

}
